src/Microsoft.AspNetCore.Sockets.Common/TextMessageBatchFormatter.cs

```diff
@@ -0,0 +1,239 @@
+        /// </summary>
+        /// <param name="data">The data containing the encoded messages to read</param>
+        /// <exception cref="FormatException">The input does not adhere to the protocol</exception>
+        /// <returns>A list of <see cref="Message"/> values decoded from the buffer</returns>
+        public static IEnumerable<Message> ReadMessages(ReadOnlySpan<byte> data)
+        {
+            if (data.Length < 1 || data[0] != 'T')
+            {
+                throw new FormatException("Missing 'T' prefix in Text Message Batch.");
+            }
+
+            int cursor = 1;
+            while (cursor < data.Length)
+            {
+                // Parse the length
+                int length = ParseLength(data, ref cursor);
+                cursor++;
+
+                // Parse the type
+                var type = ParseType(data, ref cursor);
+                cursor++;
+
+                // Read the payload
+                var payload = ParsePayload(data, length, type, ref cursor);
+                cursor++;
+
+                yield return new Message(payload, type, endOfMessage: true);
+            }
+        }
+
+        /// <summary>
+        /// Attempts to write the specified messages to the target buffer
+        /// </summary>
+        /// <param name="target">The buffer in which to encode the messages</param>
+        /// <param name="messages">The messages to encode</param>
+        /// <returns>A boolean indicating if there was enough space in the buffer to encode the message.</returns>
+        public static bool WriteMessages(Span<byte> target, IReadOnlyList<Message> messages)
+        {
+            var size = MeasureMessages(messages);
+            if(target.Length < size)
+            {
+                return false;
+            }
+
+            int cursor = 0;
+            target[cursor] = (byte)'T';
+            cursor++;
+
+            foreach(var message in messages)
+            {
+                WriteMessage(target, message, ref cursor);
+            }
+            return true;
+        }
+
+        private static void WriteMessage(Span<byte> target, Message message, ref int cursor)
+        {
+            // Write the length as a UTF-8 string
+            var len = message.Payload.Buffer.Length.ToString();
+            var buf = Encoding.UTF8.GetBytes(len);
+            buf.CopyTo(target.Slice(cursor, buf.Length));
+            cursor += buf.Length;
+
+            // Separator
+            target[cursor] = (byte)':';
+            cursor++;
+
+            // Write the type character
+            target[cursor] = GetTypeCharacter(message.Type);
+            cursor++;
+
+            // Separator
+            target[cursor] = (byte)':';
+            cursor++;
+
+            // Payload
+            if(message.Type == MessageType.Binary)
+            {
+                var payload = Encoding.UTF8.GetBytes(Convert.ToBase64String(message.Payload.Buffer.ToArray()));
```

Comment:
We have dependency conflict issues with the latest versions of packages in corefxlab. @davidfowl would have more information.